gpytorch/utils/function_factory.py

```diff
@@ -192,19 +181,12 @@ def backward(self, grad_output):
                 if gpytorch.functions.fastest:
                     quad_part = derivative_quadratic_form_factory(*covar2_args)(mat_inv_y, mat_inv_y)
 
-                    sample_matrix = torch.sign(torch.randn(gpytorch.functions.num_trace_samples, len(mu_diff)))
-
-                    def deriv_quad_form_closure(z):
-                        rhs_vec = chol_covar1.t().mv(chol_covar1.mv(z))
-                        I_minus_Tinv_M_z = z - LinearCG().solve(covar2_matmul_closure, rhs_vec)
-                        Tinv_z = LinearCG().solve(covar2_matmul_closure, z)
-                        return derivative_quadratic_form_factory(*covar2_args)(Tinv_z, I_minus_Tinv_M_z)
-
-                    for z in sample_matrix:
-                        quad_derivative = deriv_quad_form_closure(z)
-                        for i in range(len(covar2_args)):
-                            if grad_covar2_args[i] is not None:
-                                grad_covar2_args[i].add_(quad_derivative[i])
+                    sample_matrix = torch.sign(torch.randn(len(mu_diff), gpytorch.functions.num_trace_samples))
+                    rhs_vectors = chol_covar1.t().mm(chol_covar1.mm(sample_matrix))
+                    I_minus_Tinv_M_vectors = sample_matrix - LinearCG().solve(covar2_matmul_closure, rhs_vectors)
+                    Tinv_vectors = LinearCG().solve(covar2_matmul_closure, sample_matrix)
+                    grad_covar2_args = list(derivative_quadratic_form_factory(*covar2_args)(Tinv_vectors.t(),
```

Comment:
Not for this PR, but at some point we should go through and evaluate whether any of the transpose operations we do can be removed by storing matrices better as their transposes already. Doing operations with the transpose of a matrix can be quite the slowdown due to poor memory access patterns.

Reply:
Alternatively if we must transpose a matrix, it might be worth making some of them contiguous before performing a lot of operations with them.

Reply:
I think so. I am going through those transpose operations to see if anything can be simplified. I will also try to find if it possible to write a batch version for `toeplitz_matmul`.